client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,1607 @@
+	}
+}
+
+impl Default for PeerState {
+	fn default() -> PeerState {
+		PeerState::NotConnected
+	}
+}
+
+/// Side of [`ProtocolHandle`] responsible for all the logic. Currently all instances are
+/// owned by [`crate::Peerset`], but they should eventually be moved to corresponding protocols.
+#[derive(Debug)]
+pub struct ProtocolController<PeerStoreHandle> {
+	/// Set id to use when sending connect/drop requests to `Notifications`.
+	// Will likely be replaced by `ProtocolName` in the future.
+	set_id: SetId,
+	/// Receiver for messages from [`ProtocolHandle`].
+	from_handle: TracingUnboundedReceiver<Action>,
+	/// Number of occupied slots for incoming connections (not counting reserved nodes).
+	num_in: u32,
+	/// Number of occupied slots for outgoing connections (not counting reserved nodes).
+	num_out: u32,
+	/// Maximum number of slots for incoming connections (not counting reserved nodes).
+	max_in: u32,
+	/// Maximum number of slots for outgoing connections (not counting reserved nodes).
+	max_out: u32,
+	/// Connected regular nodes.
+	nodes: HashMap<PeerId, Direction>,
+	/// Reserved nodes. Should be always connected and do not occupy peer slots.
+	reserved_nodes: HashMap<PeerId, PeerState>,
+	/// Connect only to reserved nodes.
+	reserved_only: bool,
+	/// Next time to allocate slots. This is done once per second.
+	next_periodic_alloc_slots: Instant,
+	/// Outgoing channel for messages to `Notifications`.
+	to_notifications: TracingUnboundedSender<Message>,
+	/// Peerset handle for checking peer reputation values and getting connection candidates
+	/// with highest reputation.
+	peer_store: PeerStoreHandle,
+}
+
+impl<PeerStoreHandle: PeerReputationProvider> ProtocolController<PeerStoreHandle> {
+	/// Construct new [`ProtocolController`].
+	pub fn new(
+		set_id: SetId,
+		config: SetConfig,
+		to_notifications: TracingUnboundedSender<Message>,
+		peer_store: PeerStoreHandle,
+	) -> (ProtocolHandle, ProtocolController<PeerStoreHandle>) {
+		let (to_controller, from_handle) = tracing_unbounded("mpsc_protocol_controller", 10_000);
+		let handle = ProtocolHandle { to_controller };
+		let reserved_nodes =
+			config.reserved_nodes.iter().map(|p| (*p, PeerState::NotConnected)).collect();
+		let controller = ProtocolController {
+			set_id,
+			from_handle,
+			num_in: 0,
+			num_out: 0,
+			max_in: config.in_peers,
+			max_out: config.out_peers,
+			nodes: HashMap::new(),
+			reserved_nodes,
+			reserved_only: config.reserved_only,
+			next_periodic_alloc_slots: Instant::now(),
+			to_notifications,
+			peer_store,
+		};
+		(handle, controller)
+	}
+
+	/// Drive [`ProtocolController`]. This function returns when all instances of
+	/// [`ProtocolHandle`] are dropped.
+	pub async fn run(mut self) {
+		while self.next_action().await {}
+	}
+
+	/// Perform one action. Returns `true` if it should be called again.
+	///
+	/// Intended for tests only. Use `run` for driving [`ProtocolController`].
+	pub async fn next_action(&mut self) -> bool {
```

Comment:
I imagine this to be a thin wrapper over the slot allocation so having runners and `next_action()` might not be needed in the future but let's see what makes sense. For now it makes sense

Reply:
Although we still have to have some runner for listening to events from `ProtocolHandle`. We'll get back to this at some point